### PR TITLE
fix: invalid memory address or nil pointer dereference (#550)

### LINKED ISSUE #550: KongConsumer without username crashes KIC
NOTE: GitHub issues are reserved for **bug reports only**.
For anything else, please join the conversation
in Kong Nation https://discuss.konghq.com/c/kubernetes.

---

### Summary

Hey all,

I've been running into an issue getting this controller up and running. I've managed to figure out what the cause seems to be and have managed to work around it, but I think this is a genuine bug.

```
-------------------------------------------------------------------------------
Kong Ingress controller
  Release:    0.7.0
  Build:      e4605db
  Repository: git@github.com:kong/kubernetes-ingress-controller.git
  Go:         go1.13.1
-------------------------------------------------------------------------------

I0224 10:39:36.250499       1 main.go:407] Creating API client for https://100.64.0.1:443
I0224 10:39:36.256804       1 main.go:451] Running in Kubernetes Cluster version v1.15 (v1.15.6) - git (clean) commit 7015f71e75f670eb9e7ebd4b5749639d42e20079 - platform linux/amd64
I0224 10:39:36.267404       1 main.go:187] kong version: 1.4.3
I0224 10:39:36.267423       1 main.go:196] Kong datastore: off
I0224 10:39:36.582344       1 controller.go:224] starting Ingress controller
I0224 10:39:36.584853       1 status.go:201] new leader elected: kong-kong-74ddc87c5b-6v665
E0224 10:39:36.589789       1 runtime.go:73] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 180 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x142a5c0, 0x2229630)
        /go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190817020851-f2f3a405f61d/pkg/util/runtime/runtime.go:69 +0x7b
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
        /go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190817020851-f2f3a405f61d/pkg/util/runtime/runtime.go:51 +0x82
panic(0x142a5c0, 0x2229630)
        /usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/kong/kubernetes-ingress-controller/internal/ingress/controller.(*KongController).toDeckContent.func9(0x1, 0x0, 0x97)
        /kong-ingress-controller/internal/ingress/controller/kong.go:302 +0x47
sort.insertionSort_func(0xc002a999a8, 0xc002af5410, 0x0, 0x3)
        /usr/local/go/src/sort/zfuncversion.go:12 +0xb1
sort.stable_func(0xc002a999a8, 0xc002af5410, 0x3)
        /usr/local/go/src/sort/zfuncversion.go:171 +0x86
sort.SliceStable(0x137e3a0, 0xc002b106c0, 0xc002a999a8)
        /usr/local/go/src/sort/slice.go:27 +0xcb
github.com/kong/kubernetes-ingress-controller/internal/ingress/controller.(*KongController).toDeckContent(0xc0006e6000, 0xc001a52680, 0x0, 0x0, 0x0)
        /kong-ingress-controller/internal/ingress/controller/kong.go:301 +0x1272
github.com/kong/kubernetes-ingress-controller/internal/ingress/controller.(*KongController).OnUpdate(0xc0006e6000, 0xc001a52680, 0x0, 0x0)
        /kong-ingress-controller/internal/ingress/controller/kong.go:45 +0x4d
github.com/kong/kubernetes-ingress-controller/internal/ingress/controller.(*KongController).syncIngress(0xc0006e6000, 0x149a2a0, 0xc001a2df20, 0xc022c092c1, 0x8312ed838b1)
        /kong-ingress-controller/internal/ingress/controller/controller.go:117 +0x1fa
github.com/kong/kubernetes-ingress-controller/internal/ingress/task.(*Queue).worker(0xc0007c0f60)
        /kong-ingress-controller/internal/ingress/task/queue.go:111 +0x2df
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc00030b7a8)
        /go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190817020851-f2f3a405f61d/pkg/util/wait/wait.go:152 +0x5e
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc002a9bfa8, 0x3b9aca00, 0x0, 0x1, 0xc0006504e0)
        /go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190817020851-f2f3a405f61d/pkg/util/wait/wait.go:153 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(...)
        /go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190817020851-f2f3a405f61d/pkg/util/wait/wait.go:88
github.com/kong/kubernetes-ingress-controller/internal/ingress/task.(*Queue).Run(0xc0007c0f60, 0x3b9aca00, 0xc0006504e0)
        /kong-ingress-controller/internal/ingress/task/queue.go:59 +0x6b
created by github.com/kong/kubernetes-ingress-controller/internal/ingress/controller.(*KongController).Start
        /kong-ingress-controller/internal/ingress/controller/controller.go:232 +0xe8
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x11f3497]

goroutine 180 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
        /go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190817020851-f2f3a405f61d/pkg/util/runtime/runtime.go:58 +0x105
panic(0x142a5c0, 0x2229630)
        /usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/kong/kubernetes-ingress-controller/internal/ingress/controller.(*KongController).toDeckContent.func9(0x1, 0x0, 0x97)
        /kong-ingress-controller/internal/ingress/controller/kong.go:302 +0x47
sort.insertionSort_func(0xc002a999a8, 0xc002af5410, 0x0, 0x3)
        /usr/local/go/src/sort/zfuncversion.go:12 +0xb1
sort.stable_func(0xc002a999a8, 0xc002af5410, 0x3)
        /usr/local/go/src/sort/zfuncversion.go:171 +0x86
sort.SliceStable(0x137e3a0, 0xc002b106c0, 0xc002a999a8)
        /usr/local/go/src/sort/slice.go:27 +0xcb
github.com/kong/kubernetes-ingress-controller/internal/ingress/controller.(*KongController).toDeckContent(0xc0006e6000, 0xc001a52680, 0x0, 0x0, 0x0)
        /kong-ingress-controller/internal/ingress/controller/kong.go:301 +0x1272
github.com/kong/kubernetes-ingress-controller/internal/ingress/controller.(*KongController).OnUpdate(0xc0006e6000, 0xc001a52680, 0x0, 0x0)
        /kong-ingress-controller/internal/ingress/controller/kong.go:45 +0x4d
github.com/kong/kubernetes-ingress-controller/internal/ingress/controller.(*KongController).syncIngress(0xc0006e6000, 0x149a2a0, 0xc001a2df20, 0xc022c092c1, 0x8312ed838b1)
        /kong-ingress-controller/internal/ingress/controller/controller.go:117 +0x1fa
github.com/kong/kubernetes-ingress-controller/internal/ingress/task.(*Queue).worker(0xc0007c0f60)
        /kong-ingress-controller/internal/ingress/task/queue.go:111 +0x2df
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc00030b7a8)
        /go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190817020851-f2f3a405f61d/pkg/util/wait/wait.go:152 +0x5e
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc002a9bfa8, 0x3b9aca00, 0x0, 0x1, 0xc0006504e0)
        /go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190817020851-f2f3a405f61d/pkg/util/wait/wait.go:153 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(...)
        /go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190817020851-f2f3a405f61d/pkg/util/wait/wait.go:88
github.com/kong/kubernetes-ingress-controller/internal/ingress/task.(*Queue).Run(0xc0007c0f60, 0x3b9aca00, 0xc0006504e0)
        /kong-ingress-controller/internal/ingress/task/queue.go:59 +0x6b
created by github.com/kong/kubernetes-ingress-controller/internal/ingress/controller.(*KongController).Start
        /kong-ingress-controller/internal/ingress/controller/controller.go:232 +0xe8
```

The stack trace points to this LOC, which makes sense

https://github.com/Kong/kubernetes-ingress-controller/blob/master/internal/ingress/controller/kong.go#L302

`return strings.Compare(*content.Consumers[i].Username, *content.Consumers[j].Username) > 0`

I have some KongConsumers defined like follows

```yaml
apiVersion: configuration.konghq.com/v1
credentials:
- user-secret
custom_id: 123456
kind: KongConsumer
metadata:
  name: user
```

As you can see, I don't set the username property on the consumer as based on the docs, username is optional (either that or custom_id must be provided).

Version 1.2.0 of the kong/kong helm chart

### Kubernetes version

Client Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.3", GitCommit:"06ad960bfd03b39c8310aaf92d1e7c12ce618213", GitTreeState:"clean", BuildDate:"2020-02-11T18:14:22Z", GoVersion:"go1.13.6", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.6", GitCommit:"7015f71e75f670eb9e7ebd4b5749639d42e20079", GitTreeState:"clean", BuildDate:"2019-11-13T11:11:50Z", GoVersion:"go1.12.12", Compiler:"gc", Platform:"linux/amd64"}

### What happened

Kong pod fails to spin up as the ingress controller container crashes

### Expected behavior

The pod should spin up correctly

### Potential solution

Update the controller to it handles the case when KongConsumer objects don't have a username set

### PR DESCRIPTION
**Pull request by stevemcquaid**
**What this PR does / why we need it**:
This PR offers two different solutions to fix the issue in #550 arising out of unchecked dereferencing of pointers

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #550 

**Special notes for your reviewer**:
Included two implementations:
* [`50c0eb4`](https://github.com/Kong/kubernetes-ingress-controller/pull/916/commits/50c0eb4153f090d4675a4fc5e978f69b79fc0348) - Is a simple patch address the root issue of #550.

* [`6fc1797`](https://github.com/Kong/kubernetes-ingress-controller/pull/916/commits/6fc17978d218a3c959c546552efbdff6521a10cc) - Is a cleaner fix using `reflect` to dynamically select different fields within the `consumer` struct for use everywhere a pointer is dereferenced in the various sorting calls.

Let me know if the simple patch is preferred.

### CODE REVIEW

**Comment by hbagdi** [comment]
@stevemcquaid, thanks for the PR.
While this works in theory, in practice the solution here is more of a band-aid on the problem and will result in a problem.
The controller needs to ignore KongConsumer resources which do not have `username` field set. Such resources cannot be reconciled to Kong.

**Comment by stevemcquaid** [comment]
Thanks @hbagdi - I am pretty unfamiliar with the codebase, so I hope you can forgive my ignorance here :)

After a quick search, I found a difference where the admission controller validates consumer in: `internal/admission/validator.go::ValidateConsumer()` the KongConsumer is type: `github.com/kong/kubernetes-ingress-controller/pkg/apis/configuration/v1` (fields are `string`)

It looks like the ingress controller does not utilize the same object in: `github.com/kong/kubernetes-ingress-controller/internal/ingress/controller/kong.go::toDeckContent()`, it uses `github.com/kong/go-kong/kong::Consumer` (fields are `*string`)

TL;DR: From what I can gather, the admission controller uses a different struct (`configuration.KongConsumer`) than the ingress controller (kong.controller). I assume this is by design but hoping you might have additional context

**Comment by hbagdi** [comment]
Closing this since this was not the correct solution here. Thanks for the PR @stevemcquaid!
